codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/protocol/aws/DeserializeMiddleware.java

```diff
@@ -0,0 +1,219 @@
+    }
+
+    public GoWriter.Writable generate() {
+        deserialName = getMiddlewareName(operation);
+
+        return goTemplate("""
+
+            type $opName:L struct{
+            }
+
+            func (op *$opName:L) ID() string {
+                return "OperationDeserializer"
+            }
+
+            $handleSerialize:W
+
+            """,
+                MapUtils.of(
+                        "opName", deserialName,
+                        "handleSerialize", generateHandleDeserialize()
+                ));
+    }
+
+    private GoWriter.Writable generateHandleDeserialize() {
+        return goTemplate(
+                """
+
+                        func (op *$opName:L) HandleDeserialize (ctx $context:T, in $input:T, next $handler:T) (
+                        out $output:T, metadata $metadata:T, err error) {
+
+                            $body:W
+
+                            return out, metadata, nil
+                        }
+
+                        """, MapUtils.of(
+                        "context", SmithyGoDependency.CONTEXT.interfaceSymbol("Context"),
+                        "input", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("DeserializeInput"),
+                        "handler", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("DeserializeHandler"),
+                        "output", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("DeserializeOutput"),
+                        "metadata", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("Metadata"),
+                        "opName", deserialName,
+                        "body", generateHandleDeserializeBody())
+        );
+    }
+
+    private GoWriter.Writable generateHandleDeserializeBody() {
+        return goTemplate("""
+                $errors:W
+
+                $response:W
+            """,
+                MapUtils.of(
+                        "errors", handleResponseChecks(),
+                        "response", handleResponse()
+                ));
+    }
+
+
+    private GoWriter.Writable handleResponseChecks() {
+        return goTemplate("""
+
+                out, metadata, err = next.HandleDeserialize(ctx, in)
+                if err != nil {
+                    return out, metadata, err
+                }
+
+                resp, ok := out.RawResponse.($response:P)
+                if !ok {
+                    return out, metadata, $errorf:T("unexpected transport type %T", out.RawResponse)
+                }
+
+                if resp.Header.Get("smithy-protocol") != $protocol:S {
+                    return out, metadata, &$deserError:T{
+                        Err: $errorf:T(
+                            "unexpected smithy-protocol response header '%s' (HTTP status: %s)",
+                            resp.Header.Get("smithy-protocol"),
+                            resp.Status,
+                        ),
+                    }
+                }
+
+                if resp.StatusCode != 200 {
```

Comment:
This is probably going to be fine for 90% of requests, but anything at the 2xx level is technically a "success" status (e.g. 204 "No Content" still implies the request succeeded), so you want to check that the code is in that range.